File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> path = new ArrayList<>();
    int[] nums;

    public List<List<Integer>> subsets(int[] nums) {
        this.nums = nums;
        dfs(0);
        return ans;
    }

    void dfs(int start) {
        ans.add(new ArrayList<>(path));

        for (int i = start; i < nums.length; i++) {
            path.add(nums[i]);
            dfs(i + 1);
            path.removeLast();
        }
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
    vector<vector<int>> ans;
    vector<int> path;
    vector<int> nums;
    int n;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        this->nums = nums;
        this->n = nums.size();
        dfs(0);
        return ans;
    }

    void dfs(int i) {
        if (i == n) {
            ans.emplace_back(path);
            return;
        }
        dfs(i + 1);
        path.push_back(nums[i]);
        dfs(i + 1);
        path.pop_back();
    }
};